client/src/components/Pagination.tsx

```diff
@@ -72,6 +66,133 @@ export default function Pagination({
   );
 }
 
+interface CustomPaginationProps {
+  activePage: number;
+  innerClass?: string;
+  itemsCountPerPage: number;
+  maxPages?: number;
+  onChange: (pageNumber: number) => void;
+  totalItemsCount: number;
+}
+
+function CustomPagination({
+  activePage,
+  innerClass,
+  itemsCountPerPage,
+  onChange,
+  maxPages = 5,
+  totalItemsCount,
+}: CustomPaginationProps) {
+  if (itemsCountPerPage === 0 || totalItemsCount === 0 || maxPages <= 0) {
+    return null;
+  }
+  const totalPages = Math.ceil(totalItemsCount / itemsCountPerPage);
+  if (totalPages <= 1) {
+    return null;
+  }
+
+  const half = Math.floor(maxPages / 2);
+  const start = Math.max(1, activePage - half);
+  const end = Math.min(totalPages, start + maxPages - 1);
+  const startPage =
+    end === totalPages ? Math.max(1, totalPages - maxPages + 1) : start;
+
+  const pages = [];
+
+  // First page button: only show if the first page is not already shown
+  if (startPage !== 1) {
+    pages.push(
+      <PaginationElement
+        ariaLabel="first page"
+        content="⟪"
+        onClick={() => onChange(1)}
+        key="first"
+      />
+    );
+  }
+
+  // Previous page button: only show if there's a previous page
+  if (activePage > 1) {
+    pages.push(
+      <PaginationElement
+        ariaLabel="previous page"
+        content="⟨"
+        onClick={() => onChange(activePage - 1)}
+        key="prev"
+      />
+    );
+  }
+
+  // Page number buttons
+  for (let pageNumber = startPage; pageNumber <= end; pageNumber++) {
+    pages.push(
+      <PaginationElement
+        ariaLabel={`page ${pageNumber.toString()} of ${totalPages}`}
+        content={pageNumber}
+        onClick={() => onChange(pageNumber)}
+        extraClass={pageNumber === activePage ? "active" : ""}
+        key={pageNumber}
+      />
+    );
+  }
+
+  // Next page button: only show if there's a next page
+  if (activePage < totalPages) {
+    pages.push(
+      <PaginationElement
+        ariaLabel="next page"
+        content="⟩"
+        onClick={() => onChange(activePage + 1)}
+        key="next"
+      />
+    );
+  }
+
+  // Last page button: show it if the last page is not already included in the loop
+  if (end !== totalPages) {
+    pages.push(
+      <PaginationElement
+        ariaLabel="last page"
+        content="⟫"
+        onClick={() => onChange(totalPages)}
+        key="last"
+      />
+    );
+  }
+
+  return <ul className={cx(innerClass)}>{pages}</ul>;
```

Comment:
These links should be presented in a `nav` element:

```suggestion
  return <nav aria-label={ariaLabel}><ul className={innerClass}>{pages}</ul></nav>;
```

with `ariaLabel` being a prop and populated with a useful label, e.g. "data connectors pages"

Also, do not use `cx` for a single element, it is a no-op.

Reply:
Added the `<nav aria-label` part.

> Also, do not use cx for a single element, it is a no-op.

Removed in https://github.com/SwissDataScienceCenter/renku-ui/pull/3616/commits/8a126c1084f964d5fbce31ecf998a385a224e263

Wouldn't `cx` still do its job of sanitizing the string? E.G. handling `undefined`? I added that cause we have legacy v1 code using the element.

Reply:
`cx` is used to compute the correct `className` prop, it does not do cleanup or sanitation.